ingestr/src/tiktok_ads/tiktok_helpers.py

```diff
@@ -0,0 +1,125 @@
+import json
+
+import requests
+from dlt.common.time import ensure_pendulum_datetime
+from dlt.sources.helpers.requests import Client
+
+BASE_URL = "https://business-api.tiktok.com/open_api/v1.3/report/integrated/get/"
+
+
+def retry_on_limit(
+    response: requests.Response | None, exception: BaseException | None
+) -> bool:
+    if response is None:
+        return False
+    return response.status_code == 429
+
+
+def create_client() -> requests.Session:
+    return Client(
+        request_timeout=10.0,
+        raise_for_status=False,
+        retry_condition=retry_on_limit,
+        request_max_attempts=12,
+        request_backoff_factor=2,
+    ).session
+
+
+def flat_structure(items, time_zone="UTC"):
+    for item in items:
+        if "dimensions" in item:
+            for key, value in item["dimensions"].items():
+                if key == "stat_time_day":
+                    item["stat_time_day"] = ensure_pendulum_datetime(value).in_tz(
+                        time_zone
+                    )
+                elif key == "stat_time_hour":
+                    item["stat_time_hour"] = ensure_pendulum_datetime(value).in_tz(
+                        time_zone
+                    )
+                else:
+                    item[key] = value
+            del item["dimensions"]
+
+            for key, value in item["metrics"].items():
+                item[key] = value
+            del item["metrics"]
+
+    return items
+
+
+class TikTokAPI:
+    def __init__(self, access_token, time_zone):
+        self.headers = {
+            "Access-Token": access_token,
+        }
+        self.time_zone = time_zone
+
+    def fetch_pages(
+        self, advertiser_id: str, start_time, end_time, dimensions, metrics, filters
+    ) -> list:
+        data_level_mapping = {
+            "advertiser_id": "AUCTION_ADVERTISER",
+            "campaign_id": "AUCTION_CAMPAIGN",
+            "adgroup_id": "AUCTION_ADGROUP",
+        }
+
+        data_level = "AUCTION_AD"
+        for id_dimension in dimensions:
+            if id_dimension in data_level_mapping:
+                data_level = data_level_mapping[id_dimension]
+                break
+
+        all_items = []
+        current_page = 1
+        start_time = ensure_pendulum_datetime(start_time).to_date_string()
+        end_time = ensure_pendulum_datetime(end_time).to_date_string()
+
+        self.params = {
+            "advertiser_id": advertiser_id,
+            "report_type": "BASIC",
+            "data_level": data_level,
+            "start_date": start_time,
+            "end_date": end_time,
+            "page_size": 1000,
```

Comment:
is this configurable? ideally we should get it from the `--page-size` variable in ingestr